### PR TITLE
mediatek: Add support for Xiaomi Redmi Router AX6S

### PR DESCRIPTION
**Pull request by namidairo**
mediatek: Add support for Xiaomi Redmi Router AX6S
    
    Also known as the "Xiaomi Router AX3200" in western markets.
    SoC: MediaTek MT7622B
    RAM: DDR3 256 MiB (ESMT M15T2G16128A)
    Flash: SPI-NAND 128 MiB (ESMT F50L1G41LB)
    WLAN: 2.4/5 GHz 4T4R
    2.4 GHz: MediaTek MT7622B
    5 GHz: MediaTek MT7915E
    Ethernet: 4x 10/100/1000 Mbps
    Switch: MediaTek MT7531B
    LEDs/Keys: 2/2 (Internet + System LED, Mesh button + Reset pin)
    UART: Marked J1 on board VCC RX GND TX, beginning from "1". 3.3v, 115200n8
    Power: 12 VDC, 1.5 A
    
    Notes:
    Ethernet mac addresses exist in both u-boot-env and bdata
    Only u-boot-env is used here.
    
    Installation:
    1. Complete initial device setup
    2. Login to device and make requests to endpoint http://<PLACEHOLDER1><COMMAND>
    Replacing the <stok> with the session token from your login,
    and the <COMMAND> with the following commands to "unlock" the device.
    
    nvram set ssh_en=1
    nvram set uart_en=1
    nvram set boot_wait=on
    nvram commit
    sed -i '/flg_ssh.*release/ { :a; N; /fi/! ba };/return 0/d' /etc/init.d/dropbear
    (Replace <PASSWORD> below with desired password do not execute this line)
    echo -e “<PASSWORD>/n<PASSWORD>” | passwd root
    /etc/init.d/dropbear enable
    /etc/init.d/dropbear start
    
    3. Login to the device over ssh with root and the password chosen
    4. SCP the images over to /tmp
    5. 'mtd write factory.bin firmware', 'mtd write factory.bin firmware1',
    'mtd write -r rootfs overlay'
    
    Device should reboot at this point.
    Reverting to stock:
    Typical Xiaomi tftp that accepts their signed images.
    Triggered by holding reset pin on powerup.
    
    A simple implementation of this would be via dnsmasq's
    dhcp-boot option or using the vendor's (Windows only)
    recovery tool available on their website.
    
    Signed-off-by: Richard Huynh <voxlympha@gmail.com>

TODO:

- [x] Figure out why the ethernet macs aren't getting set 
- [ ] Installation testing
- [x] Double check if I need to shorten ubi partition, it's complaining and getting truncated to 0x370000
- [ ] Add to luci-app-advanced-reboot for switching.

### CODE REVIEW

**Comment by namidairo** [comment]
Just leaving this here for feedback and a general look-over. It's assured there are some blatant issues somewhere that require action.

@dangowrt Yes, I know you would rather prefer replacing u-boot, ATF, preloader etc. in order to get a better partition layout, with wear-leveling.

The installation steps in the commit message are a placeholder last resort method, as the easier option would most likely be embargoed for a little while.

Basically, the initramfs image boots, I've verified wifi comes up with the correct mac addresses, and ethernet port names have been verified to be correct.

**Comment by dangowrt** [comment]
We can do like for the Belkin/Linksys device and have non-UBI version as close as possible to stock layout and UBI variant which requires replacing the loader. Supporting this board in U-Boot should be rather trivial, adding more images to be generated for more devices to the installer is also already intended :)

File: target/linux/mediatek/image/mt7622.mk [hunk 1]
```diff
@@ -239,3 +239,19 @@ define Device/ubnt_unifi-6-lr-ubootmod
...
+	DEVICE_ALT0_VENDOR := Xiaomi
+  DEVICE_ALT0_MODEL := Router AX3200
+	DEVICE_DTS := mt7622-xiaomi-redmi-router-ax6s
+	DEVICE_DTS_DIR := ../dts
+	DEVICE_PACKAGES := kmod-mt7915e
+	KERNEL := kernel-bin | lzma
+	KERNEL_INITRAMFS_SUFFIX := -recovery.itb
+	KERNEL_INITRAMFS := kernel-bin | lzma | fit lzma $$(KDIR)/image-$$(firstword $$(DEVICE_DTS)).dtb with-initrd | pad-to 128k
+	IMAGES := sysupgrade.itb
+	IMAGE/sysupgrade.itb := append-kernel | fit lzma $$(KDIR)/image-$$(firstword $$(DEVICE_DTS)).dtb external-static-with-rootfs | pad-rootfs | append-metadata
```
Comment (dangowrt) [inline target/linux/mediatek/image/mt7622.mk:255]: Are you sure you want to have the rootfs included in the uImage.FIT for the bootloader to verify (that's what `external-static-with-rootfs` does)? Wouldn't it be better to have at least rootfs and rootfs_data in UBI?

File: target/linux/mediatek/mt7622/base-files/etc/board.d/02_network [hunk 2]
```diff
@@ -22,16 +22,40 @@ mediatek_setup_interfaces()
...
+mediatek_setup_macs()
+{
+	local board="$1"
+	local lan_mac=""
+	local wan_mac=""
+	local label_mac=""
+
+	case $board in
+	xiaomi,xiaomi-redmi-router-ax6s)
+		lan_mac=$(mtd_get_mac_ascii u-boot-env ethaddr)
```
Comment (dangowrt) [inline target/linux/mediatek/mt7622/base-files/etc/board.d/02_network:44]: Doesn't the bootloader pass down `ethaddr` to the kernel anyway to be applied for the first Ethernet interface? Ie. no need to handle `lan_mac` here (unless stock U-Boot is really weird on that board)

Comment (namidairo) [reply target/linux/mediatek/mt7622/base-files/etc/board.d/02_network:44]: Ah, I see. Aliasing ethernet0 to gmac0 gets that done. ethaddr_wan might have to be done here still though?

Comment (dangowrt) [reply target/linux/mediatek/mt7622/base-files/etc/board.d/02_network:44]: Yes, `wan` address will need to be set here.

File: target/linux/mediatek/image/mt7622.mk (hunk 1 above)
Comment (namidairo) [reply target/linux/mediatek/image/mt7622.mk:255]: Would it be better to have rootfs in the fit since firmware and firmware1 are 30M each, and then the entire ubi as overlay? I think the start of those two partitions would just end up being "hot" regardless with this layout.

Either way I'm not entirely sure of the image generation commands and their ramifications. Were you thinking something along the lines of append-ubi | append-rootfs and then ubiformat it onto the device?

I've removed that in the meantime.

File: target/linux/mediatek/image/mt7622.mk [hunk 3]
```diff
@@ -239,3 +239,19 @@ define Device/ubnt_unifi-6-lr-ubootmod
   ARTIFACT/bl31-uboot.fip := bl31-uboot ubnt_unifi-6-lr
 endef
 TARGET_DEVICES += ubnt_unifi-6-lr-ubootmod
+
+define Device/xiaomi_redmi-router-ax6s
+	DEVICE_VENDOR := Xiaomi
+	DEVICE_MODEL := Redmi Router AX6S
+	DEVICE_ALT0_VENDOR := Xiaomi
+  DEVICE_ALT0_MODEL := Router AX3200
```
Comment (adschm) [inline target/linux/mediatek/image/mt7622.mk:247]: Broken indent.

Comment (adschm) [reply target/linux/mediatek/image/mt7622.mk:247]: Please use the same indent as all of the other devices in this file.

File: target/linux/mediatek/image/mt7622.mk [hunk 4]
```diff
@@ -239,3 +239,19 @@ define Device/ubnt_unifi-6-lr-ubootmod
...
+	DEVICE_DTS := mt7622-xiaomi-redmi-router-ax6s
+	DEVICE_DTS_DIR := ../dts
+	DEVICE_PACKAGES := kmod-mt7915e
+	KERNEL := kernel-bin | lzma
+	KERNEL_INITRAMFS_SUFFIX := -recovery.itb
+	KERNEL_INITRAMFS := kernel-bin | lzma | fit lzma $$(KDIR)/image-$$(firstword $$(DEVICE_DTS)).dtb with-initrd | pad-to 128k
```
Comment (adschm) [inline target/linux/mediatek/image/mt7622.mk:253]: Please wrap after 74 chars. max.

File: target/linux/mediatek/mt7622/base-files/etc/board.d/02_network [hunk 5]
```diff
@@ -22,16 +22,40 @@ mediatek_setup_interfaces()
...
 
+mediatek_setup_macs()
+{
+	local board="$1"
+	local lan_mac=""
+	local wan_mac=""
+	local label_mac=""
+
+	case $board in
+	xiaomi,xiaomi-redmi-router-ax6s)
```
Comment (adschm) [inline target/linux/mediatek/mt7622/base-files/etc/board.d/02_network:43]: Second "xiaomi" needs to be removed.

File: target/linux/mediatek/image/mt7622.mk (hunk 4 above)
Comment (namidairo) [reply target/linux/mediatek/image/mt7622.mk]: Fixed, initramfs image didn't need padding either.

I note that there's a couple other contraventions of this in there, but that's out of scope for this patch/pr.

**Comment by namidairo** [comment]
Another thing I noticed was...
```
[    1.782026] Creating 11 MTD partitions on "1100d000.snfi":
[    1.787512] 0x000000000000-0x000000080000 : "Preloader"
[    1.793942] 0x000000080000-0x0000000c0000 : "ATF"
[    1.799317] 0x0000000c0000-0x000000140000 : "u-boot"
[    1.805345] 0x000000140000-0x000000180000 : "u-boot-env"
[    1.811329] 0x000000180000-0x0000001c0000 : "bdata"
[    1.816880] 0x0000001c0000-0x000000240000 : "factory"
[    1.822988] 0x000000240000-0x000000280000 : "crash"
[    1.828569] 0x000000280000-0x0000002c0000 : "crash_log"
[    1.834420] 0x0000002c0000-0x0000020c0000 : "firmware"
[    1.891140] 2 fit-fw partitions found on MTD device firmware
[    1.896799] Creating 2 MTD partitions on "firmware":
[    1.901756] 0x000000000000-0x0000002a0000 : "kernel"
[    1.911380] 0x0000002a0000-0x000001e00000 : "rootfs"
[    1.962943] mtd: device 10 (rootfs) set to be root filesystem
[    1.968918] 1 squashfs-split partitions found on MTD device rootfs
[    1.975092] 0x000001120000-0x000001e00000 : "rootfs_data"
[    2.002520] 0x0000020c0000-0x000003ec0000 : "firmware1"
[    2.058801] 0x000003ec0000-0x000008000000 : "ubi"
[    2.063497] mtd: partition "ubi" extends beyond the end of device "1100d000.snfi" -- size truncated to 0x3700000
```

Which was a bit odd. 128M should be 0x8000000 right? The stock firmware had a similar message, and had their partitions defined up to 0x8000000 in their dtb as well though.

**Comment by dangowrt** [comment]
In device tree MediaTek's proprietary BBT/BMT is enabled and that uses the final block(s) which are then hidden from user.
Also note that the current/new SPI-NAND driver in OpenWrt only allows enabling/disabling BMT for the whole flash, hence also BMT-messing-around in UBI area will result in funky behavior...

File: target/linux/mediatek/image/mt7622.mk (hunk 1 above)
Comment (namidairo) [reply target/linux/mediatek/image/mt7622.mk]: I was massively overthinking the image building variables. Much cleaner to just use the inherited values.

One odd thing with sysupgrade though, is nand_do_platform_check a bit wonky? It tries to use board name as path to extract the CONTROL file, but the sysupgrade tar has the device name instead? The check fails and is visible in CLI, but still allows flashing. In luci it's not even visible at all?

**Comment by kyoto44** [comment]
@namidairo, can you tell more about "Recalculate CRC32 of 0x4-0x10000 and write to 0x0-0x3
(Respect endianness)"?

**Comment by namidairo** [comment]
Oops, just realised that should be 0x4-0xFFFF.

But the below screenshot should explain what I mean.

image

**Comment by kyoto44** [comment]
@namidairo, so I should calculate these values from the original ROM, then write them to freshly built .itb initramfs image?

**Comment by remittor** [comment]
> The stock firmware had a similar message, and had their partitions defined up to 0x8000000 in their dtb as well though.

Stock DTB: https://pastebin.com/7bCw2u0D
```
                partition@3ec0000 {
                    label = "overlay";
                    reg = <0x3ec0000 0x3200000>;
                };
 
                partition@70c0000 {
                    label = "obr";
                    reg = <0x70c0000 0xf40000>;
                };
```
dmesg (stock firmware):
```
[    1.454325] [BBT] BMT.v2 is found at 0x3ff
[    1.458452] 13 ofpart partitions found on MTD device MTK-SNAND
[    1.464283] Creating 13 MTD partitions on "MTK-SNAND":
[    1.469421] 0x000000000000-0x000000020000 : "PL_Header"
[    1.475364] 0x000000020000-0x000000080000 : "Preloader"
[    1.481500] 0x000000080000-0x0000000c0000 : "ATF"
[    1.486963] 0x0000000c0000-0x000000140000 : "uboot"
[    1.492879] 0x000000140000-0x000000180000 : "Nvram"
[    1.498523] 0x000000180000-0x0000001c0000 : "Bdata"
[    1.504159] 0x0000001c0000-0x000000240000 : "Factory"
[    1.510281] 0x000000240000-0x000000280000 : "crash"
[    1.515616] mtdoops: ready 0, 1 (no erase)
[    1.519716] mtdoops: Attached to MTD device 7
[    1.524420] 0x000000280000-0x0000002c0000 : "crash_log"
[    1.530421] 0x0000002c0000-0x0000020c0000 : "firmware"
[    1.569706] 0x0000020c0000-0x000003ec0000 : "firmware1"
[    1.612623] 2 fit-fw partitions found on MTD device firmware1
[    1.618373] 0x0000020c0000-0x000002360000 : "kernel"
[    1.626766] 0x000002360000-0x000003ec0000 : "rootfs"
[    1.662921] mtd: device 12 (rootfs) set to be root filesystem
[    1.668684] 0x000003ec0000-0x0000070c0000 : "overlay"
[    1.730305] 0x0000070c0000-0x000008000000 : "obr"
[    1.735013] mtd: partition "obr" extends beyond the end of device "MTK-SNAND" -- size truncated to 0x500000
[    1.750835] mtk-snand 1100d000.snfi: [mtk_snand] probe successfully!
[    1.757621] mtk-nor 11014000.spi: unrecognized JEDEC id bytes: 00,  0,  0
[    1.764430] mtk-nor: probe of 11014000.spi failed with error -2
```

**Comment by namidairo** [comment]
Yes, this has already been discussed above in one of the threads somewhere above, it's because of the bmt. I've already made the size a little smaller to stop it complaining. I don't believe it would have really made a functional difference anyway, unless you were generating images in that size range for some reason.

File: target/linux/mediatek/dts/mt7622-xiaomi-redmi-router-ax6s.dts [hunk 6]
```diff
@@ -0,0 +1,342 @@
...
+		 * next consecutive stock partitions:
+		 * - overlay (ubi overlay)
+		 * - obr
+		 * resulting ~56MiB space for packages.
+		 */
+
+		partition@3ec0000 {
+			label = "ubi";
+			reg = <0x3ec0000 0x3700000>;
+		};
```
[the remark this answers is not in the document]
Comment (remittor) [reply target/linux/mediatek/dts/mt7622-xiaomi-redmi-router-ax6s.dts:289]: To completely disable `Dual Boot`, just run the following commands:
```
nvram set "boot_fw1=run boot_rd_img;bootm"
nvram commit
```

[the remark this answers is not in the document]
Comment (namidairo) [reply target/linux/mediatek/dts/mt7622-xiaomi-redmi-router-ax6s.dts:289]: I don't remember if tftp recovery would restore those variables, so it could potentially complicate or degrade the experience for those who reverted to stock later...

Of course, one could always just tell people to reset boot_fw1 to `boot_fw1=run boot_rd_img2;bootm` in the revert to stock instructions too.

It's a bit of a trade-off really. A couple less install/revert instructions + some vague dual-firmware resilience vs. ~30mib extra package space.

We were happy to sacrifice dual boot on ramips for the extra space (If I recall, it was quite a bit less free space regained there as well), but what about here? 

@dangowrt Opinion? I'm quite happy to go either way in regards to storage use on the stock-alike layout.

[the remark this answers is not in the document]
Comment (remittor) [reply target/linux/mediatek/dts/mt7622-xiaomi-redmi-router-ax6s.dts:289]: > I don't remember if tftp recovery would restore those variables, so it could potentially complicate or degrade the experience for those who reverted to stock later...

I have already tested this case. The bootloader after installing the firmware restores the default values (including `boot_fw1`). 
By the way, after that `ethaddr` was changed to the one that is hardcoded in bootloader (`ethaddr=00:0C:E7:11:22:33 `).

[the remark this answers is not in the document]
Comment (namidairo) [reply target/linux/mediatek/dts/mt7622-xiaomi-redmi-router-ax6s.dts:289]: Tested the tftp recovery yesterday, it's the same ip-as-hex scheme described on the AX3600 wiki page. However, it did not reset my ethaddr in the manner you described. I'll probably end up checking for it in the network script and then just offset it from the wan/label address if it happens, so things don't break with multiple units on the same network.

Or even easier yet, just decrement by 1 from the 2.4ghz radio mac address, as that results in the same mac address. (At least on my unit anyway)

**Comment by dangowrt** [comment]
After commit https://github.com/openwrt/openwrt/commit/b86452f8414d921058d800922849691a0fcfe2a6 it is possible to enable BBT/BMT for only the range in flash where we need it: to cover everything but not not the area used for UBI. See the follow up commit https://github.com/openwrt/openwrt/commit/83b14ec026d611e3c07eedd4869db08ca682d46d for an example.

**Comment by namidairo** [comment]
Rebased to current master, added the suggested BMT ranges, additionally included support for the other flash part that has been reported to appear in some units. (Hooray diversified supply chains)

Let me know if I've missed something, as I'm away from my build machine right now.

File: package/boot/uboot-envtools/files/mediatek_mt7622 [hunk 7]
```diff
@@ -38,6 +38,9 @@ buffalo,wsr-2533dhp2)
 ubnt,unifi-6-lr-ubootmod)
 	ubootenv_add_uci_config "/dev/mtd2" "0x0" "0x4000" "0x10000"
 	;;
+xiaomi,redmi-router-ax6s)
+	ubootenv_add_uci_config "/dev/mtd3" "0x0" "0x10000" "0x40000"
+	;;
```
Comment (dangowrt) [inline package/boot/uboot-envtools/files/mediatek_mt7622:43]: We usually also split adding setting for uboot-envtools into a separate commit titled `uboot-envtools: ...`, but I personally don't mind it happening here.

Comment (remittor) [reply package/boot/uboot-envtools/files/mediatek_mt7622:43]: You need to add this line: `ubootenv_add_uci_sys_config "/dev/mtd4" "0x0" "0x1000" "0x40000"`
Details here: https://github.com/openwrt/openwrt/pull/4936

Comment (namidairo) [reply package/boot/uboot-envtools/files/mediatek_mt7622:43]: I briefly had it in there before, but really couldn't find a use for it since the majority of the fields were in the usual uboot-env, and I already had access.

You're probably right it should go back in, as I believe you would be able to keep maintain access to devices after restoring back to stock.

Moreso for the Chinese-sold units, as these are coming out of the factory with telnet_en=0 in the bdata, whereas the "global" units are coming out with telnet_en=1, allegedly.

**Comment by Neustradamus** [comment]
To follow this PR, thanks to @namidairo and others :)

**Comment by mikeeq** [comment]
@namidairo there are some merge conflicts now in `target/linux/mediatek/mt7622/base-files/etc/board.d/02_network` ;)

File: target/linux/mediatek/image/mt7622.mk [hunk 8]
```diff
@@ -240,3 +240,22 @@ define Device/ubnt_unifi-6-lr-ubootmod
...
 endef
 TARGET_DEVICES += ubnt_unifi-6-lr-ubootmod
+
+define Device/xiaomi_redmi-router-ax6s
+  DEVICE_VENDOR := Xiaomi
+  DEVICE_MODEL := Redmi Router AX6S
+  DEVICE_ALT0_VENDOR := Xiaomi
+  DEVICE_ALT0_MODEL := Router AX3200
+  DEVICE_DTS := mt7622-xiaomi-redmi-router-ax6s
+  DEVICE_DTS_DIR := ../dts
```
Comment (dekomote) [inline target/linux/mediatek/image/mt7622.mk:250]: BOARD_NAME needs to be set to "xiaomi,redmi_router_ax6s", otherwise nand_do_platform_check can't find the folder inside the tarball and it doesn't pass validation. Currently sysupgrade doesn't work because of this.

```suggestion
  BOARD_NAME := xiaomi,redmi-router-ax6s
```

**Comment by dekomote** [comment]
> Why is this still doesn't merged? image 8 days uptime, works perfect so far!

I have one with 9 days and one with 26 days uptime. All working smoothly.

**Comment by dekomote** [comment]
> I thought we are waiting for @namidairo to fix sysupgrade as mentioned in [383d1bc#r820273097](https://github.com/openwrt/openwrt/commit/383d1bcc4ad7685708ffb0f983bced7483b3fd59#r820273097)...?

Sysupgrade has been fixed, he didn't mark it as resolved :)

**Review by dangowrt — APPROVED** [review]

**Comment by csrutil** [comment]
Thank you for the hard work, is there any updates here?

**Comment by edrikk** [comment]
Another vote to merge this.  Build is confirmed working - I installed on a factory fresh AX6S and it's looking all good.

**Comment by mikeeq** [comment]
@ynezz, @dangowrt I'm not sure who else we can tag to push merging this PR forward. 
It was tested multiple times on my two devices and by multiple different people on their devices and It's working like a charm.

**Comment by dangowrt** [comment]
Picked into master branch https://github.com/openwrt/openwrt/commit/4177de10df384e0106191e5b204b0aa11431fe77 and https://github.com/openwrt/openwrt/commit/9f9477b2751231d57cdd8c227149b88c93491d93

**Comment by mikeeq** [comment]
Cheers guys! :beers: Really good work everyone, especially @namidairo - thanks again!

**Comment by Neustradamus** [comment]
Thanks @namidairo for your adding and @dangowrt for merging!
